### ISSUE TITLE
only read a few columns from a csv

### ISSUE THREAD

**Issue by tdhock** [pola-rs/r-polars#267]
Hi @sorhawell 
I see you can use `n_rows` argument of  `read_csv_` to limit number of rows, but is there some way to limit which columns are read? I would like to do something like the following,
```r
>      my_file = tempfile()
> write.csv(iris, my_file, row.names=FALSE)
> data.table::fread(my_file, nrows = 2, select=1:2)
   Sepal.Length Sepal.Width
1:          5.1         3.5
2:          4.9         3.0
```
I looked at the man page, ?polars::csv_reader, and the closest mention I found was the following, but I don't see a dtype/string corresponding to "ignore this column" so I am posting an issue.
```
overwrite_dtype: (NULL is disable) named list of dtypes OR dtype-names,
          where name points to a column. Can overwrite dtypes during
          inference. Supported types so far are: name | alias | polars
          side dtype "Boolean" | "logical" => DataType::Boolean,
          "Float32" | "double" => DataType::Float32, "Float64" |
          "float64" => DataType::Float64, "Int32" | "integer" =>
          DataType::Int32, "Int64" | "integer64" => DataType::Int64,
          "Utf8" | "character" => DataType::Utf8,
```

**Comment by sorhawell** [comment]
hi @tdhock  :)

fast answer, it is possible to scan a data source into a LazyFrame and e.g. `select()` and/or `filter()` when calling `collect()`the optimized plan will push down projections and predicates to the internal csv_reader.

the `csv_reader`- function was some of the first I wrote in this projection, it is likely due to be updated.

```R
> my_file = tempfile()
> write.csv(iris, my_file, row.names=FALSE)
> 
> lf = pl$lazy_csv_reader(
+   my_file
+ )$filter(
+   pl$col("Species") == "setosa"  
+ )$select(
+   "Sepal.Length", "Sepal.Width"  
+ )
> 
> lf$describe_optimized_plan()
FAST_PROJECT: [Sepal.Length, Sepal.Width]

    CSV SCAN /var/folders/v1/b2c26lpn2yjd997jg_gn4fgc0000gn/T//Rtmp6fCkkc/file527052b4a418
    PROJECT 3/5 COLUMNS
    SELECTION: [(col("Species")) == (Utf8(setosa))]
> 
> lf$collect()
shape: (50, 2)
┌──────────────┬─────────────┐
│ Sepal.Length ┆ Sepal.Width │
│ ---          ┆ ---         │
│ f64          ┆ f64         │
╞══════════════╪═════════════╡
│ 5.1          ┆ 3.5         │
│ 4.9          ┆ 3.0         │
│ 4.7          ┆ 3.2         │
│ 4.6          ┆ 3.1         │
│ …            ┆ …           │
│ 5.1          ┆ 3.8         │
│ 4.6          ┆ 3.2         │
│ 5.3          ┆ 3.7         │
│ 5.0          ┆ 3.3         │
└──────────────┴─────────────┘
```

**Comment by tdhock** [comment]
hey thanks for the quick reply. I'm trying "first few columns and first few rows" (selection based on indices, not names) and I tried the following which did not work, is there another way to do it? (sorry for my naivete)
```r
> polars::pl$lazy_csv_reader(my_file,n_rows=2)$select(1:3)$collect()
shape: (3, 1)
┌─────┐
│     │
│ --- │
│ i32 │
╞═════╡
│ 1   │
│ 2   │
│ 3   │
└─────┘
```

**Comment by tdhock** [comment]
also I was confused by the presence of at least 3 different CSV reading functions, 

* `polars::pl$lazy_csv_reader`
* `polars::read_csv_`
* `polars::csv_reader`

can the difference please be documented (maybe under see also) on each of their man pages? or delete the ones which should be avoided? (at least delete the docs and do not export)

**Comment by sorhawell** [comment]
> I'm trying "first few columns and first few rows" (selection based on indices, not names)

I don't there a straight forward way to select columns by idx at scan.
However it is possible since 2 days ago Dec version to read schema and column names lazily so something like this should be possible

`lf$select(lf$columns[1:n_col])`

Will check tmr

**Comment by sorhawell** [comment]
> also I was confused by the presence of at least 3 different CSV reading functions,
> 
>     * `polars::pl$lazy_csv_reader`
> 
>     * `polars::read_csv_`
> 
>     * `polars::csv_reader`
> 
> 
> can the difference please be documented (maybe under see also) on each of their man pages? or delete the ones which should be avoided? (at least delete the docs and do not export)

naming convention of any I/O seems to have stabilized since in py-polars by this format
`Class.operation_dataformat`

so the naming in r-polars should also become

`.read_csv .scan_csv .write_csv .scan_parquet .sink_parquet .read_parquet`

very often internally the read-version is derived from the scan-version

**Comment by eitsupi** [comment]
The name `read_csv_()` seems quite strange to me; I think it would be better to change it to `pl_read_csv()` or `read_csv_pl()` (which is the same as the arrow package's `read_csv_arrow()`).

**Comment by vincentarelbundock** [comment]
> The name `read_csv_()` seems quite strange to me; I think it would be better to change it to `pl_read_csv()` (I believe rOpenSci recommends such a naming rule.) or `read_csv_pl()` (which is the same as the arrow package's `read_csv_arrow()`).

I agree it is weird. Do we need a top level exported function, or should there only be a `pl$read_csv`? Perhaps the latter would be more analogous to pypolars...

**Comment by sorhawell** [comment]
[I agree ;)](https://github.com/pola-rs/r-polars/pull/305/commits/16c62bbfc91bb47a52650b20bf08cc8e9d8c6b5a)

**Comment by vincentarelbundock** [comment]
Haha, yeah, I saw that 2 minutes after posting! Good change.

**Comment by sorhawell** [comment]
FYI @tdhock 

thanks to @etiennebacher  and @vincentarelbundock  it is now possible to do `[]`-indexing on lazyframe and the csv_reader has been renamed.

```R
> pl$scan_csv("deleteme.txt")[,1:2,drop=FALSE]$slice(0,2)$collect()
shape: (2, 2)
┌──────────────┬─────────────┐
│ Sepal.Length ┆ Sepal.Width │
│ ---          ┆ ---         │
│ f64          ┆ f64         │
╞══════════════╪═════════════╡
│ 5.1          ┆ 3.5         │
│ 4.9          ┆ 3.0         │
└──────────────┴─────────────┘


> pl$scan_csv("deleteme.txt")[1:2,1:2, drop=FALSE]$collect()
Error: Row selection using brackets is not supported for LazyFrames.

```

It is not currently possible to row index on LazyFrame. Not that it would be completely impossible, but to achieve same behavior as R without performing an intermediate costly $collect() will require some extra thought.
